Make Visual Studio 2019 a prerequisite to building this repo

natemcmaster opened the pull request:
We're preparing to consume some changes soon which will require Visual Studio 2019 (C# 8, updated .NET Core 3 SDKs, etc.)

Changes:

- Updated documentation in aspnet/AspNetCore
- Updated InstallVisualStudio.ps1 scripts
- Updated .sln files to require VS >= 16
- Updated CI config to use agents with VS 2019 installed

Commit:
Make Visual Studio 2019 a prerequisite for building this repo

natemcmaster commented on .azure/pipelines/jobs/iisintegration-job.yml:
```diff
@@ -3,8 +3,7 @@ jobs:
     buildArgs: "-ci -restore -build -test -projects src/Servers/IIS/**/*.csproj ${{ parameters.skipArgs }}"
-    poolName: "Hosted VS2017"
-    poolVmImage: "vs2017-win2016"
+    poolName: dotnet-external-temp-vs2019
```
@pakrym @jkotalik - what would you like to do here? There is no Microsoft-hosted pool that has VS 2019, and the agent in dotnet-external-temp-vs2019 doesn't run with admin rights.

cc @muratg

muratg replied:
Should we check with Arcade folks for their recommendation?

Is this agent pool something we created? If so, we can change it to run with admin perhaps?
Alternatively, we can clone the pool and change it and use it, right?

pakrym replied:
What are the options? I would prefer the tests running or at least to have a plan and a timeline to reenable them.

natemcmaster replied:
Options:

* Build tests in VS2019, copy bits over to a different machine for test execution (via Helix or via an 'artifact dependency' in our YML pipeline)
* Create a new different agent pool that has both VS2019 and the admin rights these tests need.
* Don't upgrade to VS2019 yet (but this is going to hold up other things we want to do, like use C# 8)

I like the first option.

pakrym replied:
Option 1 seems reasonable and if moving us in the right direction overall. Altough I don't think IIS tests are in the shape to be published and ran without requiring SDK on other machine.

natemcmaster replied:
Ok. Let's play around with some approaches. I expect this PR won't go in for a few days while we work on this. We're not exactly in a rush since we're still stabilizing Preview 2, so we have time to find something that works.

natemcmaster replied:
@pakrym is taking the first steps towards Helix testing in https://github.com/aspnet/AspNetCore/pull/7586. We'll leave disabled the IIS tests which require admin until they are ported to run in Helix.

Commit:
Merge branch 'master' into vs2019

Commit:
Update .sln files

natemcmaster commented:
Blocked on https://github.com/dotnet/core-eng/issues/5029

Commit:
Merge branch 'master' into vs2019

davidfowl commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -3,14 +3,6 @@
     Installs or updates Visual Studio on a local developer machine.
```
This doesn't seem to work:

![image](https://user-images.githubusercontent.com/95136/52224321-91385080-285c-11e9-9c12-ea0dc8fdfef6.png)

jkotalik replied:
Yeah those workloads have been removed from VS2019. We need to update korebuild.json to remove them or retarget if they have been renamed.

natemcmaster replied:
Do you happen to know the new workload names? The docs don't specify the new names https://docs.microsoft.com/en-us/visualstudio/install/workload-and-component-ids?view=vs-2019

natemcmaster replied:
I tried digging through some metadata. Can't figure out what should replace these workload names. Let me trying sending mail internally.

natemcmaster replied:
Looks like they're just gone, so no renaming required. I'm also going to update all .vcxproj to use Windows SDK 10.0.17134.0

natemcmaster replied:
Tested the installvisualstudio.ps1 script locally. Should be working correctly now.

pakrym commented:
We need a plan for https://github.com/aspnet/AspNetCore-Internal/issues/1764 as part of this.

Commit:
Merge branch 'master' into vs2019

natemcmaster commented:
/AzurePipelines run AspNetCore-ci

Commit:
Merge branch 'master' into vs2019

Commit:
Update Windows SDK to 17134

Commit:
Update docs

Commit:
Disable ANCM tests

natemcmaster commented:
> We need a plan for [aspnet/AspNetCore-Internal#1764](https://github.com/aspnet/AspNetCore-Internal/issues/1764) as part of this.

Agreed. I believe the plan is to keep running tests on TeamCity until Helix comes online (wip right now -> #6728 ). I'm updating this PR to remove the ANCM build jobs for now.

I believe the core-eng team is rolling out the updated VS2019 agents today, so we should be unblocked on this now. I'd like to merge soon, so PTAL, reviewers.

natemcmaster commented:
Build fails with an new error. I can't tell for sure, but my guess is it's a problem building the Blazor VSIX. @SteveSandersonMS @rynowak @ajaybhargavb  have you seen this before?

```
2019-02-05T23:03:00.8270212Z Unhandled Exception: System.Runtime.Serialization.SerializationException: Unable to find assembly 'Microsoft.VisualStudio.CodeAnalysis, Version=16.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
2019-02-05T23:03:00.8272220Z    at System.Runtime.Serialization.Formatters.Binary.BinaryAssemblyInfo.GetAssembly()
2019-02-05T23:03:00.8273711Z    at System.Runtime.Serialization.Formatters.Binary.ObjectReader.GetType(BinaryAssemblyInfo assemblyInfo, String name)
2019-02-05T23:03:00.8280961Z    at System.Runtime.Serialization.Formatters.Binary.ObjectMap..ctor(String objectName, String[] memberNames, BinaryTypeEnum[] binaryTypeEnumA, Object[] typeInformationA, Int32[] memberAssemIds, ObjectReader objectReader, Int32 objectId, BinaryAssemblyInfo assemblyInfo, SizedArray assemIdToAssemblyTable)
2019-02-05T23:03:00.8288479Z    at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.ReadObjectWithMapTyped(BinaryObjectWithMapTyped record)
```

https://dev.azure.com/dnceng/public/_build/results?buildId=88221&view=logs&jobId=f31c9f97-4411-58e7-49ac-fc73f645e6b6&taskId=6eab48b9-b4ce-58d4-7c0f-2d2ef2bbc0bd&lineStart=2937&lineEnd=2943&colStart=1&colEnd=103

Commit:
Merge branch 'master' into vs2019

Commit:
Capture binlog on build

natemcmaster commented:
Blocked again on https://github.com/Microsoft/msbuild/issues/4141.

@SteveSandersonMS @rynowak @ajaybhargavb - the error is a red herring. Something in MSBuild itself is failing and hiding the real error.

Commit:
Update to .NET Core SDK 3.0 Preview 2

natemcmaster commented:
I've updated this PR to use .NET Core SDK 3.0 **Preview 2** and still getting this error.

>  Unhandled Exception: System.Runtime.Serialization.SerializationException: Unable to find assembly 'Microsoft.VisualStudio.CodeAnalysis, Version=16.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.

On Linux/macOS, there is a new error: @ajaybhargavb @NTaylorMullen @pranavkm  - what is your recommendation for fixing this? My guess is happens as a result of having only _some_ projects in the repo using Microsoft.NET.Sdk.Razor as a PackageReference, and others using the one that ships with the Web SDK. Could we remove Microsoft.NET.Sdk.Razor  as a package reference completely?

> Microsoft.NET.Sdk.Razor.CodeGeneration.targets(99,5): error MSB4062: The "Microsoft.AspNetCore.Razor.Tasks.RazorTagHelper" task could not be loaded from the assembly /Users/vsts/agent/2.146.0/work/1/s/.nuget/packages/microsoft.net.sdk.razor/3.0.0-preview-19104-04/build/netstandard2.0/../../tasks/netstandard2.0/Microsoft.NET.Sdk.Razor.Tasks.dll. Assembly with same name is already loaded Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask. [/Users/vsts/agent/2.146.0/work/1/s/src/Security/samples/ClaimsTransformation/ClaimsTransformation.csproj]

Commit:
Merge branch 'master' into vs2019

pranavkm commented:
> Could we remove Microsoft.NET.Sdk.Razor as a package reference completely?

No, that becomes problematic if there are breaking changes in the SDK with corresponding runtime changes (and we've had a few rounds of these in the past). It makes sense to treat it as just as any other partner dependency and uniformly. I'll get the Security project sorted.

dougbu commented on korebuild.json:
```diff
@@ -12,7 +12,7 @@
       ],
       "includePrerelease": true,
-      "versionRange": "[15.8, 16.0)",
+      "versionRange": "[16.0,)",
```
Suggest excluding `17.0` while here.

dougbu reviewed (dismissed):
If the build succeeds :shipit:

Commit:
Merge branch 'master' into vs2019

Commit:
PR feedback

Commit:
Merge branch 'master' into vs2019

Commit:
Try dotnet-external-temp-staging

Commit:
update repo tasks dependencies

Commit:
Update Microsoft.Extensions.DependencyModel

Commit:
Add json.net to repo tasks

natemcmaster commented:
Welp, next layer of the upgrade onion -- https://github.com/NuGet/Home/issues/7706

Commit:
Revert "Add json.net to repo tasks"

This reverts commit ad70a546000a03e7ac6bb91dec11717cb2753d3f.

Commit:
Merge branch 'master' into vs2019

Commit:
Merge branch 'master' into vs2019

natemcmaster commented:
/AzurePipelines run AspNetCore-ci (Build: Windows x64/x86)

Commit:
Merge branch 'master' into vs2019

Commit:
Use Microsoft.NET.Sdk.Razor as a package consistently accross the repo

Commit:
Move helix targets

Commit:
React to changes in metadata from Microsoft.NETCore.App

Commit:
React to changes in .NET Core SDK

Commit:
Revert updates to repotask dependencies

Commit:
Merge branch 'master' into vs2019

Commit:
Update dependencies

Commit:
Update C++ toolset to v142

Workaround https://developercommunity.visualstudio.com/content/problem/434385/vs2019-preview-2-targetframeworkversion-or-platfor.html

Commit:
Require Microsoft.VisualStudio.Workload.VisualStudioExtension

Commit:
Merge branch 'master' into vs2019

natemcmaster commented:
Ok, found a workaround for MSBuild crashing (manually apply this change to msbuild.exe.config https://github.com/Microsoft/msbuild/pull/4139)

Next error: there are 3 compiler errors in C++ projects with VS2019. Not sure if this is the result of updating the platform toolset to `v142` to workaround [this bug](https://developercommunity.visualstudio.com/content/problem/434385/vs2019-preview-2-targetframeworkversion-or-platfor.html), or the result of other changes in VS2019.

Signalr Client. @BrennanConroy - any ideas why this header is missing?
```
C:\src\aspnet\AspNetCore\src\SignalR\clients\cpp\src\signalrclient\stdafx.h(21): fatal error C1083: Cannot open include
 file: 'cpprest/details/basic_types.h': No such file or directory [C:\src\aspnet\AspNetCore\src\SignalR\clients\cpp\src
\signalrclientdll\Build\VS\signalrclientdll.vcxproj]
C:\src\aspnet\AspNetCore\src\SignalR\clients\cpp\src\signalrclient\stdafx.h(21): fatal error C1083: Cannot open include
 file: 'cpprest/details/basic_types.h': No such file or directory [C:\src\aspnet\AspNetCore\src\SignalR\clients\cpp\src
\signalrclient\Build\VS\signalrclient.vcxproj]
```
ANCM - @jkotalik is this a legit bug in our code that we should fix?
```
c:\src\aspnet\aspnetcore\src\servers\iis\aspnetcoremodulev2\aspnetcore\servererrorapplication.h(32): error C26487: Don'
t return a pointer '(*pRequestHandler)' that may be invalid (lifetime.4). [C:\src\aspnet\AspNetCore\src\Servers\IIS\Asp
NetCoreModuleV2\AspNetCore\AspNetCore.vcxproj]
```

BrennanConroy commented:
Yeah, `v142` will not work with the cpprestsdk and I don't see anything in there issue tracker for adding support for it.

muratg commented:
@BrennanConroy is there a related issue you can add this comment to? If not, shouldn't we file one? Let's not stay blocked for long without an issue tracking it.

natemcmaster commented:
Ok, I've opened https://github.com/Microsoft/cpprestsdk/issues/1039. In the meantime, I'm going to try reseting back to v141 and attempt to workaround https://developercommunity.visualstudio.com/content/problem/434385/vs2019-preview-2-targetframeworkversion-or-platfor.html

Commit:
Revert to v141 toolset

Commit:
Merge branch 'master' into vs2019

Commit:
Add v141 and VC 14.16 tools to require workloads

natemcmaster commented:
Next error:

> C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(321,5): error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. 

Seems like the installation of VS on our CI machine is busted.

Commit:
Don't produce binlog

Commit:
Merge branch 'master' into vs2019

Commit:
Update template tests workarounds

Commit:
Override nuget restore path for template tests

Commit:
React to changes in .NET Core SDK

Commit:
Commit changes to package.json

Commit:
fixup! Commit changes to package.json

Commit:
Fix broken tests on macOS/Linux

Commit:
Merge branch 'master' into vs2019

natemcmaster commented:
There is no good workaround for https://developercommunity.visualstudio.com/content/problem/434385/vs2019-preview-2-targetframeworkversion-or-platfor.html, but according to the internal thread, this should be resolved in VS 2019 Preview 3. Let's wait until we have Preview 3 instead of trying to hack our way around this bug in the C++ tooling.

BrennanConroy commented:
Do we have a rough ETA for that? I would almost recommend temporarily removing the C++ client from our build until then, as I'm the only one who will be really working on it.

natemcmaster commented:
It's not just SignalR. ANCM stuff fails to compile as well. 

See the [schedule](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki?wikiVersion=GBwikiMaster&pagePath=%2FProduct%20Lifecycle%2FDev16%20Release%2FDev16%20Schedule&pageId=980) for an ETA...not sure if I can disclose this on GitHub.

BrennanConroy commented:
But ANCM was fine updating to 142 right?

natemcmaster commented:
ANCM fails with v142

> c:\src\aspnet\aspnetcore\src\servers\iis\aspnetcoremodulev2\aspnetcore\servererrorapplication.h(32): error C26487: Don't return a pointer '(*pRequestHandler)' that may be invalid (lifetime.4). [C:\src\aspnet\AspNetCore\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj]

BrennanConroy commented:
Oh, that should be an easy fix, either turn off the code analysis rules or fix the rule break. @pakrym @jkotalik

Commit:
Merge branch 'master' into vs2019

Commit:
Merge branch 'master' into vs2019

Commit:
React to renaming of VS2019 workloads

Commit:
Update min version to VS 2019 Preview 3

Commit:
Remove usage of VCToolsRedistVersion

natemcmaster commented:
I've opened https://github.com/aspnet/AspNetCore/issues/7548 and https://github.com/aspnet/AspNetCore/issues/7549 to track updating the platform toolset in a separate PR. The CI agents have been updated to include a workaround for the C++ tooling issue.

natemcmaster commented:
Next problem: when building with VS2019, the C++ projects fail due to a header which cannot be found. @jkotalik @pakrym - any ideas?
```
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file acache.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file ahutil.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file base64.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file multisz.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file multisza.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file stringa.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file stringu.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
e:\a\_work\1\s\src\servers\iis\aspnetcoremodulev2\iislib\precomp.h(10): fatal error C1083: Cannot open include file: 'atlcomcli.h': No such file or directory (compiling source file util.cpp) [E:\A\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
```

pakrym commented:
@natemcmaster I'll see if we can drop it.

Commit:
Add atl to required workloads

natemcmaster commented:
In the meantime, I'm adding "Microsoft.VisualStudio.Component.VC.v141.ATL" as a required workload.

pakrym commented:
Nope, we use quite a bit of types from it. No easy way to drop.

Commit:
Workaround Microsoft/msbuild#4150 and issue in C++ tooling

Commit:
Add --no-restore flags during CI build

Commit:
Merge branch 'master' into vs2019

natemcmaster commented:
/azp run aspnetcore-ci

natemcmaster committed:
Update ci.yml

natemcmaster committed:
Update default-build.yml

natemcmaster committed:
Update ci.yml

natemcmaster committed:
Update default-build.yml

natemcmaster committed:
Update CodeCheck.ps1

natemcmaster committed:
Update helix-test.yml

Commit:
Merge branch 'master' into vs2019

Commit:
Merge branch 'master' into vs2019

Commit:
Update build definitions

Commit:
Fix `--no-restore` in build.sh

Commit:
Attempt to workaround CodeCheck.ps1 failure which doesn't repro locally or on different agents. Possibly due to differences in the version of the PowerShell task?

Commit:
Use git diff --name-only to figure out _which_ files are failing code check

Commit:
Merge branch 'master' into vs2019

Commit:
Download git during codecheck

Commit:
Revert "Download git during codecheck"

This reverts commit 35ef92e87857fadd40c941a91ffe9db75c8e65a6.

Commit:
Debug git version

Commit:
Merge branch 'master' into vs2019

Commit:
Emit PSVersion table

Commit:
Merge branch 'master' into vs2019

Commit:
redirect stderr from git.exe to null

Commit:
Suppress first-run console output

natemcmaster commented:
Status update: we're waiting until we have more VS2019 agents on CI. We currently have 1 shared with the whole dotnet team, which is not going to be sufficient.

In the meantime, I see the code-check step still consistently fails. For some unclear reason, PowerShell is treating output to stderr as a fatal error, but only on the VS2019 agent machines. This script works just find locally (Win 10, PowerShell 5.1) and on the Hosted VS2017 pool, but fails consistently in the PR check.

```
Run git diff to check for pending changes

Summary:

   0 error(s)

##[error]git : warning: LF will be replaced by CRLF in eng/ProjectReferences.props.
##[error]At E:\A\_work\5\s\eng\scripts\CodeCheck.ps1:132 char:21
+ ... ngedFiles = git --no-pager diff --ignore-space-at-eol --name-only 2>  ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (warning: LF wil...ferences.props.:String) [], RemoteException

##[error]    + FullyQualifiedErrorId : NativeCommandError
```

Commit:
Try setting git config core.safecrlf false on CI

Commit:
Revert "Try setting git config core.safecrlf false on CI"

This reverts commit 9eb25c1a1d40595c9d90a1444dfebf33ec212786.

Commit:
Attempt to handle stderr from git better

Damn you, PowerShell

Commit:
Merge branch 'master' into vs2019

Commit:
Exclude "The file will have its original line endings in your working directory"

Commit:
Remove dead YML file

natemcmaster commented:
Resolved the git.exe problem by redirecting stderr to stdout and filtering the warning messages.

I believe this is ready for final review and ready to merge (pending CI checks, of course).

natemcmaster commented:
A bunch of test failures. It looks like they fall into 3 categories:

* HttpSys response caching is broken (cc @Tratcher any ideas?)
* MusicStore - running `dotnet publish` fails, but not error is present in logs
* MusicStore - some tests timeout trying to update the database

Commit:
Rename usages of win7-{x64,x86} to win-{x64,x86}

Tratcher commented:
@natemcmaster - HttpSys - No, nothing should have changed here. The thing most likely to destabilize those tests it is changing the OS version.

natemcmaster commented:
Yeah, that might be it. I've updated to make windows RIDs in test consistent across all projects.

Commit:
Update KoreBuild to 3.0.0-build-20190219.1

natemcmaster commented:
All required checks passed.

Eilon commented on eng/Workarounds.props:
```diff
@@ -0,0 +1,20 @@
+<!-- Use this file to workaround issues. List the issue tracking the item to fix so we can remove the workaround when the issue is resolved. -->
```
Nice!

Eilon reviewed (approved):
Looks a'ight to me!

Commit:
Merge branch 'master' into vs2019

Commit:
Apparently DOTNET_SKIP_FIRST_TIME_EXPERIENCE is not respected

Tratcher commented:
"The version of the selected solution is higher than the solution version supported by this Visual Studio release. Please try downloading the latest service pack for this Visual Studio release, or opening in a more recent version of Visual Studio, if available."

Hosting.sln
```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 16
VisualStudioVersion = 16.0.0.0
MinimumVisualStudioVersion = 16.0.28608.199
```

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 3.0
VisualStudio.16.Preview/16.0.0-pre.3.0+28608.199
Microsoft .NET Framework
Version 4.7.03190

Installed Version: Enterprise

natemcmaster commented:
Hmm, setting a specific build number of VS worked in P2. I wonder if something changed in P3. I'll reset the .sln files to 16.0.0.0.

Tratcher commented:
`MinimumVisualStudioVersion = 16.0.28606.0` seems to be the highest value that works.

petke commented:
Hi. What is the status here? I dont see any cpprestsdk.v142 nuget package.

Pilchie commented:
@jkotalik - Can you answer @petke's question?

petke commented:
@Pilchie. Sorry, I commented on the wrong project. So no need to answer me. Thanks

jkotalik commented:
No worries.
